[WIP] Feature/add card crud (mongodb will be added)

sashakurta opened the pull request:
Додав простенькі круди для карток
Додав тести до них
Додав плагін щоб був більш інформативний вивід результатів тестів

ajax-kurta-o committed:
Add CRUD for cards

ajax-kurta-o committed:
Add trigger

ajax-kurta-o committed:
Add trigger

ajax-kurta-o committed:
remove useless unit test for service

ajax-kurta-o committed:
add step to run unit tests and plugin to stdout report to console

ajax-kovtunov-o commented on .github/workflows/merge_pipeline.yml:
```diff
@@ -6,7 +6,6 @@ name: CI
   push:
-    branches: [ "master" ]
   pull_request:
     branches: [ "master" ]
```
```yaml
on:
  push:
    branches:
      - 'master'
      - 'release/**'
  pull_request:
    branches:
      - '**'
```

ajax-kovtunov-o commented on .github/workflows/merge_pipeline.yml:
```diff
@@ -34,5 +33,5 @@ jobs:
         run: ./gradlew detekt
 
-      - name: Code Coverage
+      - name: Run Test
```
Run checks

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/AnkiApplication.kt:
```diff
@@ -2,8 +2,11 @@ package io.github.anki.anki
+
 
 @SpringBootApplication
+@RestController("/")
```
why?

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/controllers/CardsController.kt:
```diff
@@ -0,0 +1,47 @@
+
+
+@RestController
+@RequestMapping("/card")
```
```kotlin
@RestController("api/v1/cards")
```

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/controllers/CardsController.kt:
```diff
@@ -0,0 +1,47 @@
+
+    @ExceptionHandler(IllegalArgumentException::class)
+    fun handleBadRequest(e: IllegalArgumentException): ResponseEntity<String> =
+        ResponseEntity(e.message, HttpStatus.BAD_REQUEST)
```
Use `RestControllerAdvice`
https://www.bezkoder.com/spring-boot-restcontrolleradvice/

<img width="647" alt="image" src="https://github.com/user-attachments/assets/36b39fae-d25f-4ee0-aa75-2d81012131a9">

ajax-kovtunov-o commented on src/test/kotlin/io/github/anki/anki/controllers/CardsControllerTest.kt:
```diff
@@ -0,0 +1,183 @@
+package io.github.anki.anki.controllers
```
```
io.github.anki.anki.controller
```

ajax-kovtunov-o commented on src/test/kotlin/io/github/anki/anki/controllers/CardsControllerTest.kt:
```diff
@@ -0,0 +1,183 @@
+    @Nested
+    @DisplayName("GET /card/{id}")
+    @TestInstance(Lifecycle.PER_CLASS)
+    inner class GetCard {
```
You don't need to create nested classes

sashakurta replied:
removed

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/models/Card.kt:
```diff
@@ -0,0 +1,13 @@
+    var description: String,
+)
+
+data class Card(
```
The models should be created for each layer

persistance layer
MongoCard

service layer
Card

controller layer
CardDto


For each layer them should be mapped

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/repository/mock/MockCardsDataSource.kt:
```diff
@@ -0,0 +1,42 @@
+import java.util.*
+
+@Repository
+class MockCardsDataSource: CardsDataSource {
```
What do you think about connecting mongo?

sashakurta replied:
https://github.com/users/sashakurta/projects/3/views/2?pane=issue&itemId=71405993 i am going to do it here
in this task only crud functionality)

ajax-kovtunov-o commented on src/main/resources/static/intex.html:
```diff
@@ -0,0 +1,10 @@
+<!DOCTYPE html>
```
Do u planing create frontend with html templates? What do u think about separated backend and separated frontend?

sashakurta replied:
remove html wil implement seperatly

ajax-kovtunov-o commented on src/test/kotlin/io/github/anki/anki/repository/mock/MockCardsDataSourceTest.kt:
```diff
@@ -0,0 +1,19 @@
+import org.junit.jupiter.api.Test
+import org.springframework.util.Assert
+
+class MockCardsDataSourceTest {
```
The Repository should be covered by integration test

ajax-kovtunov-o commented on src/test/kotlin/io/github/anki/anki/repository/mock/MockCardsDataSourceTest.kt:
```diff
@@ -0,0 +1,19 @@
+        val card: Card = mockDataSource.retrieveCard(0)
+
+        // then
+        Assert.notNull(card, "Should not be null")
```
you need some assertion lib, no one use default asserts)
for example kotest assertion lib `io.kotest:kotest-assertions-core`

usage like 
```kotlin
extension.s3Endpoint.get() shouldBe ""
```

`shouldBe` is function from kotest

sashakurta replied:
на правах реклами)?

ajax-kovtunov-o replied:
з кожного скачування отримую свою копійчину

ajax-kovtunov-o commented on src/test/kotlin/io/github/anki/anki/repository/mock/MockCardsDataSourceTest.kt:
```diff
@@ -0,0 +1,19 @@
+    private val mockDataSource = MockCardsDataSource()
+
+    @Test
+    fun `should provide some mock data`() {
```
Name tests by principle 

`should <do> when <condition>`

or 

`should <do> always`

For example `should set s3Endpoint to empty when property not set`

Cashtan54 commented on src/test/kotlin/io/github/anki/anki/controllers/CardsControllerTest.kt:
```diff
@@ -0,0 +1,183 @@
+    inner class PatchExistingBank {
+
+        @Test
+        fun `should update an existing bank`() {
```
card but not bank

sashakurta replied:
removed

ajax-kurta-o committed:
Remove base controller, rename base url, remove html static, refactor pipeline

ajax-kurta-o committed:
Add DTO objects and dto to model map
Add mongodb integration and change repository to mongo
Change docker compose file to add volume
Refactor tests
Change get to get by collection

Cashtan54 commented on .github/workflows/merge_pipeline.yml:
```diff
@@ -34,5 +26,5 @@ jobs:
         run: ./gradlew detekt
 
-      - name: Code Coverage
+      - name: Run checks
```
merge master here

Cashtan54 commented on src/main/kotlin/io/github/anki/anki/controller/CardsController.kt:
```diff
@@ -0,0 +1,51 @@
+    private val service: CardsService) {
+
+
+    @GetMapping("/collection/{collectionId}")
```
як на мене якийсь дивний урл виходить 
api/v1/cards/collections/{id}
може краще будемо повертати всі картки колекцію в геті коллекцію по айді?

ajax-kovtunov-o replied:
If you want to return cards by collection in my opinion will be more consistent to implement this functionality in collections controller.

/api/v1/collections/<id>/cards

Cashtan54 commented on src/main/kotlin/io/github/anki/anki/controller/model/dtos.kt:
```diff
@@ -0,0 +1,38 @@
+        }
+
+
+    fun toEntity(): Card {
```
maybe let`s make interface for this methods?
i mean toEntity and toModel
the same for models in service and repository

ajax-kurta-o committed:
fix linter

ajax-kovtunov-o committed:
Fix?

ajax-kurta-o committed:
Remove configuration, fix application.properties and fix 1 test

ajax-kovtunov-o commented on .github/workflows/merge_pipeline.yml:
```diff
@@ -1,28 +1,20 @@
-# This is a basic workflow to help you get started with Actions
```
NIT
I would like to rename to build.yaml

ajax-kovtunov-o commented on build.gradle.kts:
```diff
@@ -26,11 +27,12 @@ dependencies {
 	implementation("org.jetbrains.kotlin:kotlin-reflect")
-	developmentOnly("org.springframework.boot:spring-boot-devtools")
-	developmentOnly("org.springframework.boot:spring-boot-docker-compose")
+//	developmentOnly("org.springframework.boot:spring-boot-devtools")
```
Remove commented

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/controller/CardsController.kt:
```diff
@@ -0,0 +1,51 @@
+        return ResponseEntity(service.deleteCard(id), HttpStatus.OK)
+    }
+
+
```
Remove unused. Is detekt properly configured?

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/controller/model/CardDto.kt:
```diff
@@ -0,0 +1,38 @@
+import io.github.anki.anki.service.model.Card
+import org.bson.types.ObjectId
+
+
```
Remove empty line

sashakurta replied:
+

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/controller/model/CardDto.kt:
```diff
@@ -0,0 +1,38 @@
+    var cardValue: String,
+) {
+
+    private var _id: String? = null
```
It should be id

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/controller/model/CardDto.kt:
```diff
@@ -0,0 +1,38 @@
+        }
+
+
+    fun toEntity(): Card {
```
Mapper should be external function or class

The entity usually means database model, and this naming related to service layer object little bit confusing.

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/controller/model/CardDto.kt:
```diff
@@ -0,0 +1,38 @@
+
+
+    fun toEntity(): Card {
+        val entity = Card(parentCollectionId, cardValue, cardKey)
```
This objects should be immutable.

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/exception/ControllerExceptionHandler.kt:
```diff
@@ -0,0 +1,40 @@
+    @ExceptionHandler(ResourceNotFoundException::class)
+    @ResponseStatus(value = HttpStatus.NOT_FOUND)
+    fun resourceNotFoundException(ex: ResourceNotFoundException, request: WebRequest): ErrorMessage {
+        val message = ErrorMessage(
```
No any sense to create variable, just return the object

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/exception/ControllerExceptionHandler.kt:
```diff
@@ -0,0 +1,40 @@
+package io.github.anki.anki.exception
```
this shouldn't be located at exception package, move to controller

sashakurta replied:
+

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/exception/ControllerExceptionHandler.kt:
```diff
@@ -0,0 +1,40 @@
+
+    @ExceptionHandler(ResourceNotFoundException::class)
+    @ResponseStatus(value = HttpStatus.NOT_FOUND)
+    fun resourceNotFoundException(ex: ResourceNotFoundException, request: WebRequest): ErrorMessage {
```
use `org.springframework.http.ResponseEntity`

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/controller/model/CardDto.kt:
```diff
@@ -0,0 +1,38 @@
+package io.github.anki.anki.controller.model
```
rename package to `dto`

sashakurta replied:
+

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/controller/model/CardDto.kt:
```diff
@@ -0,0 +1,38 @@
+import org.bson.types.ObjectId
+
+
+class CardDto (
```
`DtoCard`? like `MongoCard`
or change mongo card instead to `CardMongo`

sashakurta replied:
+ CardDto and MongoCard

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/controller/model/CardDto.kt:
```diff
@@ -0,0 +1,38 @@
+
+    private var _id: String? = null
+
+    var id: String?
```
this should be immutable value and

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/controller/model/CardDto.kt:
```diff
@@ -0,0 +1,38 @@
+
+
+class CardDto (
+    var parentCollectionId: String,
+    var cardKey: String,
+    var cardValue: String,
```
should be immutable you also can use data class

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/controller/model/CardDto.kt:
```diff
@@ -0,0 +1,38 @@
+        return entity
+    }
+
+    fun toModel(): MongoCard {
```
This class should know anything about mongo

the mapper creates like this

```kotlin
package io.github.anki.anki.controller.dto.mapper

fun DtoCard.toMongo(): MongoCard = MongoCard.from(
  cardKey = cardKey,
  cardValue = cardValue
)
```

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/exception/ErrorMessage.kt:
```diff
@@ -0,0 +1,9 @@
+package io.github.anki.anki.exception
+
+import java.util.*
```
imports with * considered as bad practice

sashakurta replied:
+

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/repository/mongodb/CardRepository.kt:
```diff
@@ -0,0 +1,14 @@
+
+@Repository
+interface CardRepository: MongoRepository<MongoCard, String> {
+    fun findByParentCollectionId(parentCollectionId: String): List<MongoCard>
```
provide some space between methods

sashakurta replied:
+

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/repository/mongodb/model/MongoCard.kt:
```diff
@@ -0,0 +1,35 @@
+            _id = value
+        }
+
+    fun toEntity(): Card {
```
Mapper should be external function or class. Check comments from dto.

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/repository/mongodb/model/MongoCard.kt:
```diff
@@ -0,0 +1,35 @@
+        return card
+    }
+
+    fun toDto(): CardDto {
```
Mapper should be external function or class

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/repository/mongodb/model/MongoCard.kt:
```diff
@@ -0,0 +1,35 @@
+
+
+@Document(collection = "card")
+data class MongoCard(
```
Here example of model from org api, try do somthing like this
```kotlin 
@Document(collection = MongoApiPermissionsBundle.COLLECTION_NAME)
@TypeAlias(MongoApiPermissionsBundle.TYPE_ALIAS)
internal data class MongoApiPermissionsBundle(
    @Id
    override val id: ObjectId? = null,
    @CreatedDate
    override val createdAt: Instant? = null,
    @LastModifiedDate
    override val modifiedAt: Instant? = null,

    val bundleName: String? = null,
    val authorities: List<String>? = null,
    val fieldsPermissions: MongoApiFieldsPermissions? = null
) {
  companion object {
        const val COLLECTION_NAME = "permissions_bundle"
        const val TYPE_ALIAS = "MongoApiPermissionsBundle"
    }
}
```

sashakurta replied:
+

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/service/CardsService.kt:
```diff
@@ -0,0 +1,38 @@
+class CardsService(
+    private val dataSource: CardRepository) {
+
+    fun getCards(collectionId: String): List<Card> {
```
How should it works? The no any collection model present in your PR. 
I highly recommend implement functionality with relation.

### in the next PR
ONLY if one card can be associated only to one collection, then   
1a. Add to Card field named `collectionId`
Otherwise simplified way can be like this
1b. Add to Card field array named collections, that will store all collections related to this card
But if you want to search cards by collections you probably won't to iterate by all cards collection to find all related cards, then 
1c. Remove all information about collection from Card and create additional collection that can be named CardsCollections and store relations between cards and collections here. The search by collection in this case would be much faster compared to 1b

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/service/model/Card.kt:
```diff
@@ -0,0 +1,36 @@
+import io.github.anki.anki.repository.mongodb.model.MongoCard
+import org.bson.types.ObjectId
+
+class Card (
```
Immutable, without mappers, you probably already saw previous comments

ajax-kovtunov-o commented on src/test/kotlin/io/github/anki/anki/controller/CardsControllerTest.kt:
```diff
@@ -0,0 +1,198 @@
+        dataSource.deleteById(model.id.toString())
+    }
+
+//    @AfterTest
```
commented code?!) oh nice

ajax-kovtunov-o commented on src/test/kotlin/io/github/anki/anki/controller/CardsControllerTest.kt:
Provide tests for mappers, for service

If you check that database state have changed in controller test you could skip repository testing. But of course nice to have additional repository test.

sashakurta replied:
+

ajax-kurta-o committed:
Refactor tests, refactor mappings and remove get
